Adding an input folder param to dotnet pack.

livarcocc opened the pull request:
Adding an input folder param to dotnet pack. If this param is present, pack does not do a build and instead tries to get the contents of the package from the input folder. Also refactored a bunch, separating the concerns in different classes and making just program just handle the arguments and invokes the builders.

livarcocc commented:
@piotrpMSFT @brthor

livarcocc commented:
#340

piotrpMSFT commented on src/Microsoft.DotNet.Tools.Pack/InputOutputPathsCalculator.cs:
```diff
@@ -0,0 +1,58 @@
+            }
+        }        
+
+        public InputOutputPathsCalculator(Project project, string inputPath, string outputPath, string configuration)
```
We have an extension method on ProjectContext for this. This type also mixes responsibilities between "GetOutputPath" and "outputPath ?? GetOutputPath". I'm not sure I see the value.

livarcocc replied:
What is this extension method? What does it do? 

I just moved code around from Program.cs to here. What this class is doing is pretty much what was being done in program.cs plus figuring out the right InputPath, which will depend on wether an input option is passed (we use it), or not (in which case we use OutputPath).

piotrpMSFT replied:
I think we should be more specific with names... input path ==> compile artifacts path. output path ==> package artifacts path?

livarcocc replied:
Done!

piotrpMSFT commented on src/Microsoft.DotNet.Tools.Pack/InputOutputPathsCalculator.cs:
```diff
@@ -0,0 +1,58 @@
+
+        public string InputPathForContext(ProjectContext context)
+        {
+            return OutputPathSet ? OutputPath : Path.Combine(InputPath, context.TargetFramework.GetTwoDigitShortFolderName());
```
why is InputPathForContext returning an OutputPath? These names are very confusing.

livarcocc replied:
Actually, this was a good catch peter, this was a copy/paste problem. I had a OutputPathForContext before, which I copied and modified for this. I fixed it to be InputPathSet and InputPath.

davidfowl commented:
We should make sure this aligns with anything nuget pack does or will do. We can keep this alice if we fork the behavior. 
 
PS: refactoring as part of new feature work like this is pretty painful to review

piotrpMSFT commented on src/Microsoft.DotNet.Tools.Pack/PackagesGenerator.cs:
```diff
@@ -0,0 +1,41 @@
+using Microsoft.DotNet.Tools.Compiler;
+
+namespace Microsoft.DotNet.Tools.Pack
+{
```
Feels like this should be static. Is an instance interesting here?

piotrpMSFT commented on src/Microsoft.DotNet.Tools.Pack/Program.cs:
```diff
@@ -24,29 +23,30 @@ public static int Main(string[] args)
             app.Description = "Packager for the .NET Platform";
             app.HelpOption("-h|--help");
 
+            var input = app.Option("-i|--input <INPUT_DIR>", "Directory from where the assets to be packages are going to be picked up", CommandOptionType.SingleValue);
```
should input, output, and project all be suffixed with Path? This would be consistent with usage in other classes, and helps reason about these special strings.

livarcocc replied:
This is how it was before, for output and project at least, I don't want to change it.

piotrpMSFT commented on src/Microsoft.DotNet.Tools.Pack/ProjectBuilder.cs:
```diff
@@ -0,0 +1,68 @@
+
+namespace Microsoft.DotNet.Tools.Pack
+{
+    internal class ProjectBuilder
```
same here... why instance?

piotrpMSFT replied:
The Builder verb sounds like StringBuilder... a thing that itself does building. This is a thing that invokes the Build command. BuildProjectCommand.Execute()?

livarcocc replied:
Done.

piotrpMSFT commented on src/Microsoft.DotNet.Tools.Pack/PackagesGenerator.cs:
```diff
@@ -0,0 +1,41 @@
+            var symbolsPackageGenerator = 
+                new SymbolPackageGenerator(project, _configuration, _inputOutputPathsCalculator);
+
+            return mainPackageGenerator.BuildPackage(_contexts, packDiagnostics) &&
```
I think we should move towards exception-based flow for return codes.

piotrpMSFT commented on src/Microsoft.DotNet.Tools.Pack/InputOutputPathsCalculator.cs:
```diff
@@ -0,0 +1,58 @@
+
+namespace Microsoft.DotNet.Tools.Pack
+{
+    public class InputOutputPathsCalculator
```
ArtifactPathsCalculator? InputOutputPathsCalculator seems funny to me. I think this implies more of a relationship than exists between them. This just tells us where to find input artifacts and to put output artifacts.

livarcocc replied:
Done.

piotrpMSFT commented:
Looks good, aside from some nits on naming. To @davidfowl's point, consider NuGet's param names:
http://docs.nuget.org/consume/command-line-reference#pack-command

davidfowl commented:
Can you show how this will be used? O assume we have a scenario for this right now.

livarcocc commented:
@davidfowl Signing. After the publish step, we can point the signing service to the bin folder, have it sign the dlls and exes we care about and after that we can have pack generate nuget packages from the signed binaries by passing the input folder.

livarcocc commented:
@dotnet-bot Test Windows_NT Debug Build please
@dotnet-bot Test Windows_NT Release Build please

piotrpMSFT commented:
@livarcocc looks like --input is being passed to DNX?

livarcocc commented:
@piotrpMSFT I noticed. I didn't realize that I actually sent this file in PR and when I changed it to basepath, I left it out of the commit. I modified it already to basepath and I am running a local build before pushing it.

livarcocc commented:
@dotnet-bot test OSX Debug Build please

livarcocc commented:
@dotnet-bot Test OSX Debug Build please
